services/billing-azure/src/main/java/com/epam/dlab/billing/azure/AzureBillableResourcesService.java

```diff
@@ -160,62 +161,56 @@ private String getConfigurationSettingValueOrEmpty(String key) {
 	private Set<AzureDlabBillableResource> getEdgeAndStorageAccount() {
 		Set<AzureDlabBillableResource> billableResources = new HashSet<>();
 
-		try {
-
-			final FindIterable<Document> prjDocuments = mongoDbBillingClient.getDatabase()
-					.getCollection("Projects").find();
-			final List<Document> edges = StreamSupport.stream(prjDocuments.spliterator(), false)
-					.flatMap(d -> ((List<Document>) d.get("endpoints")).stream())
-					.map(d -> (Document) d.get("edgeInfo"))
-					.collect(Collectors.toList());
-			List<EdgeInfoAzure> edgeInfoList = objectMapper.readValue(
-					objectMapper.writeValueAsString(edges),
-					new com.fasterxml.jackson.core.type.TypeReference<List<EdgeInfoAzure>>() {
-					});
-
-			if (edgeInfoList != null && !edgeInfoList.isEmpty()) {
-				for (EdgeInfoAzure edgeInfoAzure : edgeInfoList) {
-					billableResources.addAll(getEdgeAndStorageAccount(edgeInfoAzure));
-				}
+		Map<String, List<Document>> projectEndpoints = StreamSupport.stream(mongoDbBillingClient.getDatabase()
+				.getCollection("Projects").find().spliterator(), false)
+				.collect(Collectors.toMap(key -> key.getString("name").toLowerCase(),
+						value -> (List<Document>) value.get("endpoints")));
+
+		projectEndpoints.forEach((key, value) -> value.forEach(endpoint -> {
+			try {
+				billableResources.addAll(getEdgeAndStorageAccount(key, objectMapper.readValue(
+						objectMapper.writeValueAsString(endpoint.get("edgeInfo")),
+						new com.fasterxml.jackson.core.type.TypeReference<EdgeInfoAzure>() {
+						})));
+			} catch (IOException ex) {
+				log.error("Error during preparation of billable resources", ex);
```

Comment:
it is better to use stream api here than foreach inside foreach:
billableResources.addAll(
l,stream()
.map (d -> EdgeInfoAzure)
.collect(toList()))